src/jit/lir.cpp

```diff
@@ -1407,6 +1407,92 @@ LIR::ReadOnlyRange LIR::Range::GetRangeOfOperandTrees(GenTree* root, bool* isClo
 #ifdef DEBUG
 
 //------------------------------------------------------------------------
+// CheckLclVarSemanticsHelper checks lclVar semantics.
+//
+// Specifically, ensure that an unaliasable lclVar is not redefined between the
+// point at which a use appears in linear order and the point at which that use is consumed by its user.
+// This ensures that it is always safe to treat a lclVar use as happening at the user (rather than at
+// the lclVar node).
+class CheckLclVarSemanticsHelper
+{
+public:
+    //------------------------------------------------------------------------
+    // CheckLclVarSemanticsHelper constructor: Init arguments for the helper.
+    //
+    // This needs unusedDefs because unused lclVar reads may otherwise appear as outstanding reads
+    // and produce false indications that a write to a lclVar occurs while outstanding reads of that lclVar
+    // exist.
+    //
+    // Arguments:
+    //    compiler - A compiler context.
+    //    range - a range to do the check.
+    //    unusedDefs - map of defs that do no have users.
+    //
+    CheckLclVarSemanticsHelper(Compiler*         compiler,
+                               const LIR::Range* range,
+                               SmallHashTable<GenTreePtr, bool, 32U>& unusedDefs)
+        : compiler(compiler), range(range), unusedDefs(unusedDefs), unconsumedLclVarReads(compiler)
+    {
+    }
+
+    //------------------------------------------------------------------------
+    // Check: do the check.
+    // Return Value:
+    //    'true' if the Local variables semantics for the specified range is legal.
+    bool Check()
+    {
+        for (GenTreePtr node : *range)
+        {
+            ConsumeNodeOperands(node);
+
+            AliasSet::NodeInfo nodeInfo(compiler, node);
+            if (nodeInfo.IsLclVarRead() && !unusedDefs.Contains(node))
+            {
+                int count = 0;
+                unconsumedLclVarReads.TryGetValue(nodeInfo.LclNum(), &count);
+                unconsumedLclVarReads.AddOrUpdate(nodeInfo.LclNum(), count + 1);
+            }
+
+            // If this node is a lclVar write, it must be to a lclVar that does not have an outstanding read.
+            assert(!nodeInfo.IsLclVarWrite() || !unconsumedLclVarReads.Contains(nodeInfo.LclNum()));
+        }
+
+        return true;
+    }
+
+private:
+    //------------------------------------------------------------------------
+    // ConsumeNodeOperands: mark the node's operands as consumed.
```

Comment:
I don't think this should talk about consuming node operands. The term "consume" has specific meaning in the backend, and it appears that this is using that term to mean something different.

Reply:
Il example was added.

Reply:
I used the same naming as it was before in `unconsumedLclVarReads`. Do we want rename `unconsumedLclVarReads` as well?

Reply:
> Do we want rename unconsumedLclVarReads as well?

Yes, I think so. That is not as bad, because it is not claiming to consume them, which has implications for liveness updates, etc. But I think it's bad practice to refer to consuming at all in this context.

Reply:
I think what we do is actually  liveness analysis here.
I can't find a better name for it, but if you have one I will rename `unconsumedLclVarReads` and `ConsumeNodeOperands` to that.

Reply:
We aren't quite doing liveness here, though we are doing something somewhat related. What we are trying to do in ensure that given a node `t0` that represents a use of a lclVar `V0` and its using node `tX`, there are no defs of `V0` between `t0` and `tX`. This is a requirement of our model wherein lclVar uses occur at the user rather than at the lclVar node.

I think that we could use `unusedLclVarReads` and `UseNodeOperands` in place of `unconsumedLclVarReads` and `ConsumeNodeOperands`.

Reply:
Fixed.